Fixes for bugs resuming request fibers when connection send loop exits

gavin-norman-sociomantic opened the pull request:
I'd suggest we hold off merging this until @david-eckardt-sociomantic has a look.

Commit:
Add missing import for debug(SwarmConn)

Commit:
Fix order of params doc

Commit:
Do not attempt to resume running fiber when shutting down node request

Previously, the connection would call Request.error, which is an alias for
RequestOnConn.resumeFiber. This method calls RequestOnConn.resumeFiber_,
which the node overrides to assert if the fiber is already running.

In the case when the connection is being shut down due to an error, one
request fiber is usually still running. This will be the request which
resumed the connection's send loop by calling SendLoop.registerForSending.
This fiber (as it is running) cannot simply be resumed -- this would cause
the assertion mentioned above to fire.

This commit adds a check that only resumes non-running fibers in this
circumstance.

nemanja-boric-sociomantic commented on src/swarm/neo/node/Connection.d:
```diff
@@ -256,8 +256,6 @@ class Connection: ConnectionBase
     override protected void notifyShutdown ( RequestId id, Exception e )
```
re: commit message - perhaps it's worth saying why there's there a running request fiber while the send loop's fiber is also running?

gavin-norman-sociomantic replied:
Is this not enough?
>This will be the request which resumed the connection's send loop by calling SendLoop.registerForSending.

nemanja-boric-sociomantic replied:
Yeah, this should be enough :+1:

david-eckardt-sociomantic commented on src/swarm/neo/connection/ConnectionBase.d:
```diff
@@ -156,6 +189,11 @@ abstract class ConnectionBase: ISelectClient
+                    // If the send loop exited, notify the calling fiber. (See
+                    // note on error handling, above.)
+                    if ( !this.loop_started )
+                        throw send_loop_exited_ex;
```
Hm, wouldn't it make more sense to throw the exception which cause the send loop to exit?

gavin-norman-sociomantic replied:
Yeah... We'd have to dup it I guess...? By the time we get to this point, the exception no longer exists.

gavin-norman-sociomantic replied:
Updated. (Retested with https://github.com/sociomantic-tsunami/dhtnode/issues/13.)

david-eckardt-sociomantic replied:
For the record, the code has already been updated: Making `send_loop_exited_ex` non-static, assigning the exception caught in `fiberMethod` and rethrowing it here is sufficient. The exception doesn't need to be dupped.

david-eckardt-sociomantic replied:
Two more things:
- `throw this.send_loop_exited_ex;`
- Reset `this.send_loop_exited_ex = null` after throwing (e.g. in a `scope (exit)` clause).

gavin-norman-sociomantic replied:
Updated. (Also added an assert that the exception is non-null before throwing it.)

Commit:
When connection send loop exits, inform calling fiber of error

The send loop fiber only exits in the case of errors on the connection.
When this occurs, we need to make sure the request fiber which resumed the
send loop (i.e. by calling SendLoop.registerForSending) is made aware of
the error. This is achieved by throwing after the call to SendLoop.resume
returns, if the send loop is no longer running.

gavin-norman-sociomantic commented:
Anyone feel like merging?
